src/interpreter/mod.rs

```diff
@@ -314,38 +317,86 @@ impl<R: RealNumberInternalTrait, E: IEnvironment<R>> Interpreter<R, E> {
         Self::eval_expression(&expression, &self.env)
     }
 
+    #[allow(unused_assignments)]
     pub fn apply_procedure<'a>(
-        procedure: &Procedure<R, E>,
-        evaluated_args: ArgVec<R, E>,
+        initial_procedure_call: &ProcedureCall,
+        initial_env: &Rc<RefCell<E>>,
     ) -> Result<Value<R, E>> {
-        Ok(match &procedure {
-            Procedure::Buildin(BuildinProcedure { pointer, .. }) => pointer(evaluated_args)?,
-            Procedure::User(SchemeProcedure(formals, definitions, expressions), closure) => {
-                Self::apply_scheme_procedure(
-                    &formals,
-                    &definitions,
-                    &expressions,
-                    closure.clone(),
-                    evaluated_args,
-                )?
+        let mut procedure_call = initial_procedure_call;
+        let mut current_procedure_call = None;
+        let mut env = initial_env;
+        let mut current_env: Option<Rc<RefCell<E>>> = None;
+        loop {
+            let first = Self::eval_expression(&procedure_call.procedure_expr, env)?;
+            let evaluated_args_result: Result<ArgVec<R, E>> = procedure_call
+                .arguments
+                .iter()
+                .map(|arg| Self::eval_expression(arg, env))
+                .collect();
+            let (procedure, evaluated_args) = match first {
+                Value::Procedure(procedure) => (procedure, evaluated_args_result?),
+                _ => logic_error!("expect a procedure here"),
+            };
+            if let Some(current_env) = &current_env {
+                current_env.borrow_mut().clear_local_definitions();
+            }
+            match procedure {
+                Procedure::Buildin(BuildinProcedure { pointer, .. }) => {
+                    break pointer(evaluated_args)
+                }
+                Procedure::User(SchemeProcedure(formals, definitions, expressions), closure) => {
+                    let local_env = current_env
```

Comment:
I think it should be renamed as sub_closure, not envs, as those envs are built when a procedure is defined, not when it been called.